ldap/servers/slapd/slapi-memberof.c

```diff
@@ -0,0 +1,1280 @@
+#if MEMBEROF_CACHE_DEBUG
+                    slapi_log_err(SLAPI_LOG_ERR, "linked_attributes", "merge_ancestors: skip (already present) %s\n", slapi_value_get_string(sval_ndn));
+#endif
+                    slapi_value_free(&sval_dn);
+                    slapi_value_free(&sval_ndn);
+                    slapi_value_free(&sval_nsuniqueid);
+                }
+                slapi_sdn_free(&val_sdn);
+            }
+        }
+        hint = slapi_valueset_next_value(v1_groupvals, hint, &sval);
+        hint_nsuniqueid = slapi_valueset_next_value(v1_nsuniqueidvals, hint_nsuniqueid, &sval_2);
+    }
+}
+
+
+/*
+ * Return 1 if the entry is in the scope.
+ * For MODRDN the caller should check both the preop
+ * and postop entries.  If we are moving out of, or
+ * into scope, we should process it.
+ */
+static int
+Xmemberof_entry_in_scope(Slapi_MemberOfConfig *config, Slapi_DN *sdn)
+{
+    if (config->entryScopeExcludeSubtrees) {
+        int i = 0;
+
+        /* check the excludes */
+        while (config->entryScopeExcludeSubtrees[i]) {
+            if (slapi_sdn_issuffix(sdn, config->entryScopeExcludeSubtrees[i])) {
+                return 0;
+            }
+            i++;
+        }
+    }
+    if (config->entryScopes) {
+        int i = 0;
+
+        /* check the excludes */
+        while (config->entryScopes[i]) {
+            if (slapi_sdn_issuffix(sdn, config->entryScopes[i])) {
+                return 1;
+            }
+            i++;
+        }
+        return 0;
+    }
+
+    return 1;
+}
+
+/* memberof_get_groups_callback()
+ *
+ * Callback to perform work of memberof_get_groups()
+ */
+int
+Xmemberof_get_groups_callback(Slapi_Entry *e, void *callback_data)
+{
+    Slapi_DN *group_sdn = slapi_entry_get_sdn(e);
+    char *group_ndn = slapi_entry_get_ndn(e);
+    char *group_dn = slapi_entry_get_dn(e);
+    const char *group_nsuniqueid = slapi_entry_get_uniqueid(e);
+    Slapi_Value *group_ndn_val = 0;
+    Slapi_Value *group_dn_val = 0;
+    Slapi_Value *group_nsuniqueid_val = 0;
+    Slapi_Value *already_seen_ndn_val = 0;
+    Slapi_ValueSet *groupvals = *((Xmemberof_get_groups_data *)callback_data)->groupvals;
+    Slapi_ValueSet *group_norm_vals = *((Xmemberof_get_groups_data *)callback_data)->group_norm_vals;
+    Slapi_ValueSet *group_nsuniqueid_vals = *((Xmemberof_get_groups_data *)callback_data)->nsuniqueidvals;
+    Slapi_ValueSet *already_seen_ndn_vals = *((Xmemberof_get_groups_data *)callback_data)->already_seen_ndn_vals;
+    Slapi_MemberOfConfig *config = ((Xmemberof_get_groups_data *)callback_data)->config;
+    int rc = 0;
+
+    if (slapi_is_shutting_down()) {
+        rc = -1;
+        goto bail;
+    }
+    
+    if (config->maxgroups_reached) {
+        rc = -1;
+        goto bail;
+    }
+
+    if (!groupvals || !group_norm_vals) {
+        slapi_log_err(SLAPI_LOG_ERR, "linked_attributes",
+                      "memberof_get_groups_callback - NULL groupvals or group_norm_vals\n");
```

Comment:
Again, "linked_attributes", copy and paste error?  I'm seeing a lot of these logging functions using "linked_attributes", I won't tag them all but please look for them.

Reply:
slapi_memberof computes memberships that I named (bad wording) linked_attributes. I understand that it is confusing, should I rather name the module 'slapi_memberof' ?

Reply:
Yes, use slapi_memberof, and in the message to the function name.  In this case your function name starts with a "X" so you should update the function name in the logging as well.

slapi_log_err(SLAPI_LOG_ERR, "slapi_memberof",
                      "Xmemberof_get_groups_callback - NULL groupvals or group_norm_vals\n");

Of course if rename the function to not start with "X" then the logging text will need to be updated.